compiler/llvm/clangUtil.cpp

```diff
@@ -4578,6 +4578,69 @@ static void saveIrToBcFileIfNeeded(const std::string& filename) {
   }
 }
 
+#ifdef HAVE_LLVM
+static llvm::MDNode* getMetadataContainingAttr(llvm::MDNode* MD, llvm::StringRef attr) {
+  for (unsigned i = 1, e = MD->getNumOperands(); i < e; ++i) {
+    if(auto md = llvm::dyn_cast<llvm::MDNode>(MD->getOperand(i))) {
+      if(auto S = llvm::dyn_cast<llvm::MDString>(md->getOperand(0))) {
+        if(S->getString() == attr) return md;
+      }
+    }
+  }
+  return nullptr;
+}
+static unsigned getBinaryMetadataValue(llvm::MDNode* MD) {
+  if (MD && MD->getNumOperands() == 2) {
+    unsigned Count =
+        llvm::mdconst::extract<ConstantInt>(MD->getOperand(1))->getZExtValue();
+    return Count;
+  }
+  return 0;
+}
+
+static void checkLoopsAssertVectorize() {
+  GenInfo* info = gGenInfo;
+  INT_ASSERT(info);
+
+  // identify all loops
+  for (const auto& F : *info->module) {
+    for (const auto& BB: F) {
+      // if the BB is well-formed, we only care about the last instruction
+      // extract from it the MD_loop, which will return something if its a BR
+      // if its not well formed, the terminator will return nullptr
+      if(auto I = BB.getTerminator()) {
+        if(auto MD = I->getMetadata(LLVMContext::MD_loop)) {
+          auto MD_assertvectorized =
+            getMetadataContainingAttr(MD, "chpl.loop.assertvectorized");
+          auto hasAssertVectorized = getBinaryMetadataValue(MD_assertvectorized) == 1;
+          if(hasAssertVectorized) {
+            auto MD_isvectorize = getMetadataContainingAttr(MD, "llvm.loop.isvectorized");
+            auto isVectorized = getBinaryMetadataValue(MD_isvectorize) == 1;
+            if(!isVectorized) {
+              const char* astr_funcCName = astr(F.getName().str());
+              auto funcIt = info->functionCNameAstrToSymbol.find(astr_funcCName);
+              FnSymbol* fn = nullptr;
+              if (funcIt != info->functionCNameAstrToSymbol.end()) {
+                fn = funcIt->second;
+              }
+              int lineno = -1;
+              if(I->getDebugLoc().get()) {
+                lineno = I->getDebugLoc().getLine();
+              }
+              else if(fn != nullptr) {
+                lineno = fn->linenum();
+              }
+              // USR_WARN should handle the case of a nullptr
+              USR_WARN(fn, "loop on line %d was marked 'assertVectorized' but did not vectorize", lineno);
```

Comment:
You may have told me this before, but I expect that passing `fn` should decorate the warning with the location of `fn`. Similarly for any other AST node. But is the problem here that we are far away from our AST and don't have proper line numbers? This is reminiscent of poor `-g` implementation issues with LLVM backend.

Reply:
Yes `fn` will decorate with the location of `fn`. But in the case that `-g` is used, we can query the DILocation metadata for the specific loop. I can specialize this warning to only have the extra part about the loop if there was debug information...

I also considered embedding line information in the attribute itself, but that would require extra specialization for this metadata in `cg-CForLoop`